support/kwok/kwok.go

```diff
@@ -0,0 +1,264 @@
+		if err := k.installKwokCtl(e, osVersion, archVersion); err != nil {
+			return err
+		}
+	}
+	if e.Prog().Avail("kwok") == "" {
+		log.V(4).Infof(`kwok not found, installing version @%s`, kwokVersion)
+		if err := k.installKwok(e, osVersion, archVersion); err != nil {
+			return err
+		}
+	}
+
+	// PATH may already be set to include $PATH:/usr/local/bin so we don't need to.
+	if kwokCtlPath := e.Prog().Avail("kwokctl"); kwokCtlPath != "" {
+		log.V(4).Info("Installed kwokctl at", kwokCtlPath)
+		// PATH may already be set to include $PATH:/usr/local/bin so we don't need to.
+		if kwokPath := e.Prog().Avail("kwok"); kwokPath != "" {
+			log.V(4).Info("Installed kwok at", kwokPath)
+			return nil
+		}
+	}
+
+	p := e.RunProc("ls $GOPATH/bin")
+	if p.Err() != nil {
+		return fmt.Errorf("failed to install kwok: %s", p.Err())
+	}
+
+	p = e.RunProc("echo $PATH:$GOPATH/bin")
+	if p.Err() != nil {
+		return fmt.Errorf("failed to install kwok: %s", p.Err())
+	}
+
+	log.V(4).Info(`Setting path to include $GOPATH/bin:`, p.Result())
+	e.SetEnv("PATH", p.Result())
+
+	if kwokCtlPath := e.Prog().Avail("kwokctl"); kwokCtlPath != "" {
+		log.V(4).Info("Installed kwokctl at", kwokCtlPath)
+		// PATH may already be set to include $PATH:/usr/local/bin so we don't need to.
+		if kwokPath := e.Prog().Avail("kwok"); kwokPath != "" {
+			log.V(4).Info("Installed kwok at", kwokPath)
+			return nil
+		}
+	}
+	return fmt.Errorf("kwok not available even after installation")
+}
+
+func (k *Cluster) installKwokCtl(e *gexe.Echo, osVersion string, archVersion string) error {
+
+	dname, err := os.MkdirTemp("", "kwok-install")
+	if err != nil {
+		return fmt.Errorf("failed to install kwokctl: %s", err)
+	}
+	defer os.RemoveAll(dname)
+
+	installKwokCtlCmd := fmt.Sprintf("wget -O /%s/kwokctl -c https://github.com/kubernetes-sigs/kwok/releases/download/%s/kwokctl-%s-%s", dname, kwokVersion, osVersion, archVersion)
+	log.V(4).Infof("%s", installKwokCtlCmd)
+	p := e.RunProc(installKwokCtlCmd)
+	if p.Err() != nil {
+		return fmt.Errorf("failed to install kwokctl: %s %s", installKwokCtlCmd, p.Err())
+	}
+
+	if !p.IsSuccess() || p.ExitCode() != 0 {
+		return fmt.Errorf("failed to install kwokctl: %s", p.Result())
+	}
+	p = e.RunProc(fmt.Sprintf("chmod +x /%s/kwokctl", dname))
+	if p.Err() != nil {
+		return fmt.Errorf("failed to install kwokctl: %s", p.Err())
+	}
+
+	p = e.RunProc(fmt.Sprintf("mv /%s/kwokctl $GOPATH/bin", dname))
+	if p.Err() != nil {
+		return fmt.Errorf("failed to install kwokctl: %s", p.Err())
+	}
+	return nil
+}
+
+func (k *Cluster) installKwok(e *gexe.Echo, osVersion string, archVersion string) error {
+
+	dname, err := os.MkdirTemp("", "kwok-install")
+	if err != nil {
+		return fmt.Errorf("failed to install kwokctl: %s", err)
+	}
+	defer os.RemoveAll(dname)
+	installKwokCmd := fmt.Sprintf("wget -O /%s/kwok -c https://github.com/kubernetes-sigs/kwok/releases/download/%s/kwok-%s-%s", dname, kwokVersion, osVersion, archVersion)
```

Comment:
Looks like we can `go install` the tools

```
❯ go install sigs.k8s.io/kwok/cmd/kwok@v0.1.1
go: downloading k8s.io/apimachinery v0.26.0
go: downloading github.com/containerd/go-cni v1.1.7
go: downloading github.com/containernetworking/plugins v1.1.1
go: downloading github.com/wzshiming/cron v0.2.1
go: downloading k8s.io/api v0.26.0
go: downloading github.com/wzshiming/cmux v0.3.2
go: downloading k8s.io/apiserver v0.26.0
go: downloading golang.org/x/exp v0.0.0-20221126150942-6ab00d035af9
go: downloading github.com/wzshiming/ctc v1.2.3
go: downloading golang.org/x/time v0.2.0
go: downloading github.com/itchyny/gojq v0.12.11
go: downloading github.com/prometheus/common v0.37.0
go: downloading github.com/google/gofuzz v1.1.0
go: downloading k8s.io/kube-openapi v0.0.0-20221012153701-172d655c2280
go: downloading github.com/google/gnostic v0.5.7-v3refs
go: downloading golang.org/x/oauth2 v0.2.0
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2
go: downloading github.com/wzshiming/llrb v0.2.1
go: downloading github.com/wzshiming/trie v0.1.1
go: downloading github.com/go-openapi/swag v0.19.14
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/containernetworking/cni v1.1.1
go: downloading github.com/wzshiming/winseq v0.0.0-20200112104235-db357dc107ae
go: downloading github.com/itchyny/timefmt-go v0.1.5
go: downloading k8s.io/component-base v0.26.0
go: downloading github.com/google/uuid v1.1.2

❯ kwok -h
kwok is a tool for simulate thousands of fake kubelets

Usage:
  kwok [command] [flags]

Flags:
      --cidr string                                        CIDR of the pod ip (default "10.0.0.1/24")
  -c, --config stringArray                                 config path (default [/var/home/bpratt/.kwok/kwok.yaml])
      --disregard-status-with-annotation-selector string   All node/pod status excluding the ones that match the annotation selector will be watched and managed.
      --disregard-status-with-label-selector string        All node/pod status excluding the ones that match the label selector will be watched and managed.
      --experimental-enable-cni                            Experimental support for getting pod ip from CNI, for CNI-related components, Only works with Linux
  -h, --help                                               help for kwok
      --kubeconfig string                                  Path to the kubeconfig file to use (default "/var/home/bpratt/.kube/config")
      --manage-all-nodes                                   All nodes will be watched and managed. It's conflicted with manage-nodes-with-annotation-selector and manage-nodes-with-label-selector.
      --manage-nodes-with-annotation-selector string       Nodes that match the annotation selector will be watched and managed. It's conflicted with manage-all-nodes.
      --manage-nodes-with-label-selector string            Nodes that match the label selector will be watched and managed. It's conflicted with manage-all-nodes.
      --master string                                      Server is the address of the kubernetes cluster
      --node-ip string                                     IP of the node
      --node-name string                                   Name of the node
      --node-port int                                      Port of the node
      --server-address string                              Address to expose health and metrics on
      --tls-cert-file string                               File containing the default x509 Certificate for HTTPS
      --tls-private-key-file string                        File containing the default x509 private key matching --tls-cert-file
  -v, --v int                                              number for the log level verbosity
      --version                                            version for kwok

```

Reply:
+1 to doing it via the `go install`. We do the same with `kind` as well

Reply:
Thanks, tried doing that. looks like `kwokctl create cluster` command when downloaded this way.  Noticed that it does not parse version correctly this way and throws `unknown version`
```
reeta@Reetas-MacBook-Pro ~ % go install sigs.k8s.io/kwok/cmd/kwok@v0.1.1
reeta@Reetas-MacBook-Pro ~ %
reeta@Reetas-MacBook-Pro ~ %
reeta@Reetas-MacBook-Pro ~ % kwokctl --version
kwokctl version unknown
reeta@Reetas-MacBook-Pro ~ % kwokctl create cluster --name abc
Creating cluster                                                                                                                   cluster=abc
Pull image                                                                                    cluster=abc image="registry.k8s.io/etcd:unknown"
Error response from daemon: manifest for registry.k8s.io/etcd:unknown not found: manifest unknown: Failed to fetch "unknown"
ERROR Failed to setup config                                         cluster=abc err="docker pull registry.k8s.io/etcd:unknown: exit status 1"
Cluster is cleaned up                                                                                                              cluster=abc
ERROR Execute exit                                                               err="docker pull registry.k8s.io/etcd:unknown: exit status 1"
reeta@Reetas-MacBook-Pro ~ %

```

Reply:
also, created an issue https://github.com/kubernetes-sigs/kwok/issues/520 for this

Reply:
https://github.com/kubernetes-sigs/kwok/issues/520 got fixed in the upstream, I have changed to using `go install`